src/SAWScript/Crucible/LLVM/X86.hs

```diff
@@ -0,0 +1,514 @@
+  ptr <- C.LLVM.doPtrAddOffset sym mem' base =<< W4.bvLit sym knownNat szInt
+  (ptr,) <$> C.LLVM.doStore sym mem' ptr (C.LLVM.LLVMPointerRepr $ knownNat @64)
+    (C.LLVM.bitvectorType 8) C.LLVM.noAlignment fresh
+
+-- | Process a crucible_alloc statement, allocating the requested memory and
+-- associating a pointer to that memory with the appropriate index.
+executeAllocation ::
+  Sym ->
+  ( Map MS.AllocIndex Ptr {- ^ Associates each previous AllocIndex with the corresponding allocation -}
+  , Mem
+  ) ->
+  (MS.AllocIndex, LLVMAllocSpec) {- ^ crucible_alloc statement -} ->
+  IO (Map MS.AllocIndex Ptr, Mem)
+executeAllocation sym (env, mem) (i, LLVMAllocSpec mut _memTy sz loc) = do
+  let ?ptrWidth = knownNat @64
+  sz' <- liftIO $ W4.bvLit sym knownNat $ C.LLVM.bytesToInteger sz
+  (ptr, mem') <- C.LLVM.doMalloc sym C.LLVM.HeapAlloc mut
+    (show $ W4.plSourceLoc loc) mem sz' C.LLVM.noAlignment
+  pure (Map.insert i ptr env, mem')
+
+-- | Process a crucible_points_to statement, writing some SetupValue to a pointer.
+executePointsTo ::
+  Sym ->
+  LLVMCrucibleContext LLVMArch ->
+  Map MS.AllocIndex Ptr {- ^ Associates each AllocIndex with the corresponding allocation -} ->
+  Map MS.AllocIndex LLVMAllocSpec {- ^ Associates each AllocIndex with its specification -} ->
+  Map MS.AllocIndex C.LLVM.Ident {- ^ Associates each AllocIndex with its name -} ->
+  Mem ->
+  LLVMPointsTo LLVMArch {- ^ crucible_points_to statement from the precondition -} ->
+  IO Mem
+executePointsTo sym cc env tyenv nameEnv mem (LLVMPointsTo _ tptr tval) = do
+  let ?ptrWidth = knownNat @64
+  ptr <- C.LLVM.unpackMemValue sym (C.LLVM.LLVMPointerRepr $ knownNat @64)
+    =<< resolveSetupVal cc mem env tyenv Map.empty tptr
+  val <- resolveSetupVal cc mem env tyenv Map.empty tval
+  storTy <- C.LLVM.toStorableType =<< typeOfSetupValue cc tyenv nameEnv tval
+  C.LLVM.storeConstRaw sym mem ptr storTy C.LLVM.noAlignment val
+
+-- | Write each SetupValue passed to crucible_execute_func to the appropriate
+-- x86_64 register from the calling convention.
+setArgs ::
+  Sym ->
+  LLVMCrucibleContext LLVMArch ->
+  Map MS.AllocIndex Ptr {- ^ Associates each AllocIndex with the corresponding allocation -} ->
+  Map MS.AllocIndex LLVMAllocSpec {- ^ Associates each AllocIndex with its specification -} ->
+  Map MS.AllocIndex C.LLVM.Ident {- ^ Associates each AllocIndex with its name -} ->
+  Mem ->
+  Regs ->
+  [MS.SetupValue LLVM] {- ^ Arguments passed to crucible_execute_func -} ->
+  IO Regs
+setArgs sym cc env tyenv nameEnv mem regs args
+  | length args > length argRegs = fail "More arguments than would fit into registers"
+  | otherwise = foldlM setRegSetupValue regs $ zip argRegs args
+  where
+    argRegs :: [Register]
+    argRegs = [Macaw.RDI, Macaw.RSI, Macaw.RDX, Macaw.RCX, Macaw.R8, Macaw.R9]
+    setRegSetupValue rs (reg, sval) = do
+      let ?ptrWidth = knownNat @64
+      val <- C.LLVM.unpackMemValue sym (C.LLVM.LLVMPointerRepr $ knownNat @64)
```

Comment:
just double-checking, what happens if the setup value `sval` is a 32bit int?

Reply:
Previously, this was causing a Crucible panic in `unpackMemValue`. It should fail more gracefully now. We could implicitly extend values shorter than 64 bits here, but I'm unsure if that's the correct thing to do. For now, if we need a 32-bit argument we can pad it with `zero #` in Cryptol.